teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,285 @@
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+  - [References](#references)
+<!-- /toc -->
+
+## Summary
+
+[TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md) introduces the Artifacts API for the use case of artifacts (backed to a Persistent Volume) that are produced and consumed within the provenance. After more reflection and discussions, we think that it is only part of the use case of the artifacts API. It does not cover the use case of artifacts like `images`, `packages`, `source code` as discussed in [this feature request](https://github.com/tektoncd/pipeline/issues/6326). We want to be able to handle different artifact types using the same API. With `StepActions` now in play, that design needs to be updated to include `StepActions`. While the Artifact API design is still undergoing early design and brainstorming, one aspect that has general agreement is around reporting of the artifact provenance data to the `TaskRun` `Status`. We therefore think that it will be nice to split the design of Artifacts support into multiple phases and incrementally add features as needed. For now, we have identified two phases to completely capture the requirements and use cases captured in [TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md) and [this feature request](https://github.com/tektoncd/pipeline/issues/6326). 
+
+This TEP proposes to define the Tekton Artifacts Provenance data structure that `Steps` can output (Phase 1). It also shows how this information is carried into the `TaskRun Status`. Furthermore, it also proposes how Tekton Chains could parse the `uri` and insert the provenance data into appropriate sections of the `SLSA` Provenance payload.
+
+## Motivation
+
+As defined in the [feature request](https://github.com/tektoncd/pipeline/issues/6326), the reporing of metadata of artifacts in Tekton has grown organically and is redundant and brittle. The structure of results make extremely challenging for a remote Task to produce artifacts of unknown length (e.g. a `ko Task` that produces images from source code does not know beforehand how many images will be produced). There is a need for users to be able to surface provenance output without being limited by the result spec.
+
+This works aims to overcome these limitations.
+
+### Goals
+#### Phase 1
+
+- Provide a mechanism for surfacing artifact provenance data to the `TaskRun` status.
+- Improve the current mechanism of collecting provenance by Tekton Chains that relies on type hinting.
+
+
+#### Phase 2
+- Provide a mechanism for making the artifact provenance data of a `TaskRun` available to downstream `TaskRuns` through variables.
+- Provide an API for users to declare upfront artifacts consumed and produced by `Tasks`.
+- Provide a set of upload and download `StepActions` for different storages.
```

Comment:
My team has been working on a stop-gap solution until trusted artifacts are in place. We created (experimental!) a [repository](https://github.com/redhat-appstudio/build-trusted-artifacts) for this. It doesn't provide a `StepAction`. Instead, it provides a container image that you can use in a step. (Adjusting to a `StepAction` seems like the a natural iteration.)

I mention it here because, hopefully, this helps with the implementation of this TEP.

We took an approach where the same image could be used to store/use artifacts from different storage backends. Today, that's just a `file`, which may or may not be backed by a PVC. `oci` is likely the next one we'll add.